HelpSource/Guides/WritingUGens.schelp

```diff
@@ -7,52 +7,210 @@ SuperCollider has a vast library of unit generators that can be assembled in unl
+}
+::
+
+The method code::hasObservableEffect:: must return code::true:: or code::false:: depending on whether the UGen alters global state, which could mean:
 list::
-## It is courteous to leave a note in your class's help file that identifies it as a pseudo-UGen.
-## To avoid confusion, pseudo-UGens should not inherit from the base class link::Classes/UGen::.
-## A SynthDef can only have two link::Classes/LocalIn::/link::Classes/LocalOut:: pairings — one for control rate and one
-for audio rate. Putting LocalIn and LocalOut in a pseudo-UGen is a bad idea, as it could interfere with other feedback
-loops in the SynthDef.
+## outputting to a bus,
+## writing to a buffer,
+## sending an OSC message,
+## freeing the synth (done action).
 ::
 
-section:: Basics of UGens
+Generally, if you would describe the UGen as 'setting' or 'writing' to something, this should be true.
 
-A (real) UGen needs two components: a plugin for the server, and a class for the language. The class goes in an ordinary
-code::*.sc:: file, and defines the interface for your UGen in the language. This class is generally just a few lines of
-code that ultimately call the class method link::Classes/UGen#-multiNew::.
+If a UGen does not indicate that it has an observable effect and is not connected to another UGen which does, it may be removed from the synthdef.
+This is called 'dead code elimination'.
+
+For UGens that have a done action, a helper method is provided which takes the index of the 'done action' input.
+
+code::
+MyUGen : UGen {
+    hasObservableEffect { ^this.implHasObservableEffectViaDoneAction(2) }
+    *ar { |arg1, arg2, doneAction| ^super.multiNew(\audio, arg1, arg2, doneAction) }
+//index:   0     1     2
+}
+::
+
+
+The default if no method is supplied is true to maintain backwards compatibility, despite most UGens actually not having observable effects.
+
+subsection:: Can Be Replaced By Identical Call
+
+code::
+MyUGen : UGen {
+    canBeReplacedByIdenticalCall { ^true }
+    *ar { |arg1, arg2| ^super.multiNew(\audio, arg1, arg2) }
+    *kr { |arg1, arg2| ^super.multiNew(\control, arg1, arg2) }
+}
+::
+
+The method code::canBeReplacedByIdenticalCall:: must return true of false depending on whether the UGen can be replaced by calls with identical inputs.
+
+Most UGens fall into this category, exceptions include:
+list::
+## noise ugens, as each instance should create unique noise,
+## and ugens that add to some shared state, link::Classes/Out:: is an example of this as it sums into the bus.
+::
+
+UGens that read from something are also replaceable with identical inputs as their position respective to other UGens that might write to the resource are considered.
+
+The default if no method is supplied is false to maintain backwards compatibility, despite most UGens being able to opt into this.
+
+subsection:: UGen Resource Manager
+
+code::
+MyBufRd : MultiOutUGen {
+	resourceManagers { ^[BufferConnectionStrategy] }
+	bufferAccessType { ^\read }
```

Comment:
ah and one more thing (sorry!): the instance methods should come after the class methods. Also when someone looks up the class, the class method interface will mostly be more relevant (the rest is for most cases implementation detail).